### ISSUE TITLE
Prediction using the model of the best echo

### ISSUE THREAD

**Issue by kitman0804** [apache/incubator-mxnet#3148]
If the best validation score occurs in epoch 25 (`num_epoch=50` was set), how can I do a prediction using the model at that epoch?

**Comment by albertyou2** [comment]
Try to use these codes : 
#-*- coding: UTF-8 -*-
from __future__ import division

import find_mxnet  
import mxnet as mx  
import logging  
import argparse  
import os, sys  
import train_model  
import numpy as np  
import cv2

parser = argparse.ArgumentParser(description='load model trained before and give prediction for .rec file')
parser.add_argument('--prefix', type=str, default='./jd_100_vgg' ,
                    help='load model structure by name which has a prefix')
parser.add_argument('--iter', type=int, default=0,
                    help='load model weight from iteration')
parser.add_argument('--shape', type=int, default=100,
                    help='data shape of .rec files')
parser.add_argument('--rgb', type=int, default=3,
                    help='color , 1 for gray,3 for color image')
parser.add_argument('--data-file', type=str, default='pred4_100.rec',
                    help='prediction .rec file path')
parser.add_argument('--min-file', type=str, default='pred4_100.min',
                    help='prediction images mean file')
parser.add_argument('--list-file', type=str, default='pred4.lst',
                    help='prediction images mean file')
args = parser.parse_args()

prefix     = args.prefix
iteration  = args.iter
model_load = mx.model.FeedForward.load(prefix, iteration)  
data_shape = (args.rgb, args.shape,args.shape)  
  
# 数据准备  batch_size = 1.    
pred = mx.io.ImageRecordIter(  
        path_imgrec = args.data_file,
        mean_img    = args.min_file,  
        rand_crop   = False,  
        rand_mirror = False,  
        data_shape  = data_shape,  
        batch_size  = 1)  
  
[prob, data, label_o] = model_load.predict(pred, return_data=True)

**Comment by kitman0804** [comment]
@albertyou2 Would your mind explaining a bit what your code doing? Thanks.

**Comment by kitman0804** [comment]
`def NeuralNet1(num_hidden=2400, activation_fun="relu"):
    data = mx.symbol.Variable('data')
    fc1  = mx.symbol.FullyConnected(data = data, name='fc1', num_hidden=num_hidden)
    act1 = mx.symbol.Activation(data = fc1, name='relu1', act_type=activation_fun)
    fc2  = mx.symbol.FullyConnected(data = act1, name='fc2', num_hidden=12)
    return mx.symbol.LogisticRegressionOutput(data = fc2, name = 'softmax')`

`nn1600_relu = mx.model.FeedForward(
    symbol=NeuralNet1(1600, "relu"), 
    num_epoch=1, learning_rate=0.05, wd=0.0001, momentum=0.9)
nn1600_relu.fit(X=itertrain, eval_metric=log_loss, eval_data=itervalid)`

The codes above gave me:

2016-08-26 21:59:34,103 Node[0] Start training with [cpu(0)]
2016-08-26 22:00:30,016 Node[0] Epoch[0] Resetting Data Iterator
2016-08-26 22:00:30,017 Node[0] Epoch[0] Time cost=55.910
2016-08-26 22:00:31,967 Node[0] Epoch[0] Validation-log_loss=2.406623

`log_loss(ytrain[ivalid], 
         nn1600_relu.predict(mx.io.NDArrayIter(xtrain[ivalid, :].toarray(), batch_size=batch_size)))`

While this code gave me 2.406555684486317 which is slightly different from 2.406623. Is there any reasons?

**Comment by albertyou2** [comment]
prefix = 'inception-bn'  ''' model prefix indicates the prefix of name of your saved network while training '''
iteration = 10               ''' epoch number indicate which  epoch's weight  you want to use(maybe the best epoch )''' 

model_load = mx.model.FeedForward.load(prefix, iteration,ctx=mx.gpu(0))    '''  load the network and its weights ,use gpu0 '''

data_shape = (3, 224,224) ''' your date shape'''

''' '''
pred = mx.io.ImageRecordIter(
path_imgrec = 'the_test_rec_file_path.rec',
mean_img =  'the_mean_file_path.min',
rand_crop = False,
rand_mirror = False,
data_shape = data_shape,
batch_size = 1)

''' prediction '''
[prob, data, label_o] = model_load.predict(pred, return_data=True)

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
